### PR TITLE
Write a reactive CSV parser

### PR DESCRIPTION
**Pull request by jypma**
...because I got bored.

Includes reading and writing, and read/write marshalling support.

### CODE REVIEW

File: ts-reaktive-marshal-akka/src/main/java/com/tradeshift/reaktive/csv/CsvParser.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+                setHandlers(in, out, new AbstractInOutHandler() {
+                    @Override
+                    public void onPull() {
+                        pull(in);
+                    }
+                    
+                    @Override
+                    public void onPush() throws CharacterCodingException {
+                        String s = buf + grab(in);
+                        System.out.println("<< " + s);
```
Comment (alar17) [inline ts-reaktive-marshal-akka/src/main/java/com/tradeshift/reaktive/csv/CsvParser.java]: You might want to get rid of the "system.out.println"s in this class.

File: ts-reaktive-marshal-akka/src/main/java/com/tradeshift/reaktive/csv/CsvParser.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+                        }
+                        completeStage();
+                    };
+                });
+            }
+            
+            private String parseAndEmit(String s) {
+                int p = (s.charAt(0) == BOM ? 1 : 0);
+                while (p < s.length()) {
+                    System.out.println("[" + p + "] " + state);
```
Comment (alar17) [inline ts-reaktive-marshal-akka/src/main/java/com/tradeshift/reaktive/csv/CsvParser.java]: This one as well

File: ts-reaktive-marshal-akka/src/main/java/com/tradeshift/reaktive/csv/CsvParser.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+                                if (p > s.length()) {
+                                    return "";
+                                }
+                            } else if (s.charAt(p) == settings.getSeparator()) {
+                                // empty field
+                                emit(out, CsvEvent.endValue());
+                                p++;
+                            } else if (s.charAt(p) == '\n') {
+                                // end of record
+                                emit(out, CsvEvent.endValue());
```
Comment (alar17) [inline ts-reaktive-marshal-akka/src/main/java/com/tradeshift/reaktive/csv/CsvParser.java]: In the case of having a csv record line this `"aaa";"bbb";"ccc"; CRLF ` I assume that the code will insert two "endValue" while parsing. I think you should either throw an error or avoid inserting the second one.

File: ts-reaktive-marshal-akka/src/main/java/com/tradeshift/reaktive/csv/CsvEvent.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
+package com.tradeshift.reaktive.csv;
+
+/**
+ * Represents a single eveny in a CSV stream.
```
Comment (alar17) [inline ts-reaktive-marshal-akka/src/main/java/com/tradeshift/reaktive/csv/CsvEvent.java]: Typo

File: ts-reaktive-marshal-akka/src/main/java/com/tradeshift/reaktive/csv/CsvParser.java (hunk 3 above)
Comment (jypma) [inline ts-reaktive-marshal-akka/src/main/java/com/tradeshift/reaktive/csv/CsvParser.java]: why? your example string has an empty column, hence I'd argue the two `endValue` are exactly how that should be represented.

Comment (alar17) [inline ts-reaktive-marshal-akka/src/main/java/com/tradeshift/reaktive/csv/CsvParser.java]: According to the RFC documentation it's a MUST NOT condition:

> The last field in the
       record must not be followed by a comma.  For example:
       aaa,bbb,ccc

Comment (jypma) [inline ts-reaktive-marshal-akka/src/main/java/com/tradeshift/reaktive/csv/CsvParser.java]: Sure. But I'd still rather do something than blow up, given the high variance in CSV. Adding capability to "jump over" empty final columns would require look-ahead logic that we don't have currently... is there really a big risk in simply outputting what's in the file in the first place?

Comment (alar17) [inline ts-reaktive-marshal-akka/src/main/java/com/tradeshift/reaktive/csv/CsvParser.java]: Fair enough. I don't think that it's a big deal. It's OK this way

**Comment by jypma** [comment]
@alar17 could you have another look?

File: ts-reaktive-csv/src/main/java/com/tradeshift/reaktive/csv/CsvTextAggregator.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                        } else {
+                            emitBuf();
+                            emit(out, event);
+                        }
+                    }
+                    
+                    public void onUpstreamFinish() throws Exception {
+                        emitBuf();
+                        completeStage();
+                    };
```
Comment (alar17) [inline ts-reaktive-csv/src/main/java/com/tradeshift/reaktive/csv/CsvTextAggregator.java]: Unnecessary semicolon

File: ts-reaktive-csv/src/test/java/com/tradeshift/reaktive/csv/marshal/CsvProtocolSpec.java [hunk 6]
```diff
@@ -2,12 +2,16 @@
 
 import static com.tradeshift.reaktive.csv.marshal.CsvProtocol.column;
 import static com.tradeshift.reaktive.csv.marshal.CsvProtocol.csv;
+import static com.tradeshift.reaktive.marshal.Protocol.option;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.forgerock.cuppa.Cuppa.describe;
 import static org.forgerock.cuppa.Cuppa.it;
+import static javaslang.control.Option.*;
```
Comment (alar17) [inline ts-reaktive-csv/src/test/java/com/tradeshift/reaktive/csv/marshal/CsvProtocolSpec.java]: Import *

File: ts-reaktive-csv/src/test/java/com/tradeshift/reaktive/csv/marshal/OptionDTO.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
+package com.tradeshift.reaktive.csv.marshal;
+
+import javaslang.control.Option;
+
+public class OptionDTO {
+    final String s;
```
Comment (alar17) [inline ts-reaktive-csv/src/test/java/com/tradeshift/reaktive/csv/marshal/OptionDTO.java]: make these private as this class is not a inner class anymore. Also consider refactoring them to a more readable names

File: ts-reaktive-csv/src/test/java/com/tradeshift/reaktive/csv/marshal/DTO.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
+package com.tradeshift.reaktive.csv.marshal;
+
+public class DTO {
+    final String s;
+    final Integer i;
```
Comment (alar17) [inline ts-reaktive-csv/src/test/java/com/tradeshift/reaktive/csv/marshal/DTO.java]: private

**Comment by jypma** [comment]
@alar17 I've addressed the final comments.

File: ts-reaktive-csv/src/main/java/com/tradeshift/reaktive/csv/CharsetDecoderFlow.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+                            throw new IllegalArgumentException("Stray bytes at end of input that could not be decoded: " + buf);
+                        }
+                        completeStage();
+                    }
+                    
+                    private void deliver(ByteBuffer bytes) throws CharacterCodingException {
+                        CharBuffer chars = CharBuffer.allocate(bytes.limit());
+                        CoderResult result = decoder.decode(bytes, chars, false);
+                        if (result.isOverflow()) {
+                            failStage(new IllegalArgumentException("Incoming bytes decoded into more characters. Huh?"));
```
Comment (jypma) [inline ts-reaktive-csv/src/main/java/com/tradeshift/reaktive/csv/CharsetDecoderFlow.java]: I think it's appropriate. It's definitely a `Huh?` :-)

**Pull request merged**